text/0000-index-assign.md

```diff
@@ -0,0 +1,325 @@
+- Feature Name: indexed_assignment
+- Start Date: 2015-05-19
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add a `IndexAssign` trait that allows overloading "indexed assignment" expressions like `a[b] = c`.
+
+# Motivation
+
+Let users define syntactic sugar for operations like these:
+
+- Inserting a *new* key-value pair into a map
+
+``` rust
+let mut map = HashMap::new();
+map[key] = value;  // equivalent to `map.insert(key, value);`
+```
+
+- Setting each element of a "slice" to some value
+
+(slice, as in a fraction of a collection, which may not necessarily be stored in contiguous memory)
+
+``` rust
+let mut matrix = { .. };
+
+// Set each element of the second row of `matrix` to `1`
```

Comment:
This sugar is common in languages aimed at numerical/scientific computing see [NumPy], [Julia] and [Matlab]/Octave. It would be helpful for libraries that want to compete in the same space to be able to provide a similar syntax to ease adoption/porting.

I don't think that this form of sugar will become common in other areas/mainstream Rust.

[NumPy]: http://wiki.scipy.org/Tentative_NumPy_Tutorial#head-864862d3f2bb4c32f04260fac61eb4ef34788c4c
[Julia]: http://docs.julialang.org/en/release-0.3/manual/arrays/#assignment
[Matlab]: http://www.mathworks.com/company/newsletters/articles/matrix-indexing-in-matlab.html